make-all.sh

```diff
@@ -1,93 +1,149 @@
-    exit 1
+if [[ `uname -s` == "MINGW"* ]]; then
+    is_windows=y
+elif [[ `uname -s` == "Darwin" ]]; then
+    is_mac=y
+else
+    is_linux=y
 fi
-if [[ ! `which python` ]]; then
-    echo "[-] Unable to find Python, is it installed?" 1>&2
-    exit 1
+
+if [[ "$is_windows" == "y" ]]; then
+    make='nmake -f Makefile.win'
+else
+    make=make
+fi
+
+python=python
+
+if [[ "$is_windows" == "y" ]]; then
+    if ! link.exe --help 2>&1 | grep -i Microsoft > /dev/null; then
+        export PATH=$(echo "$PATH" | tr ':' '\n' | grep HostX64):$PATH
+    fi
+    if ! link.exe --help 2>&1 | grep -i Microsoft > /dev/null; then
+        echo "[-] Couldn't find VS 64-bit link.exe. Make sure it's in your path!"
+        exit 1
+    fi
+else
+    if [[ ! `which cmake` ]]; then
+        echo "[-] Unable to find CMake, is it installed?" 1>&2
+        exit 1
+    fi
+    if [[ $(which python3) ]]; then
+        python=python3
+    elif [[ ! `which $python` ]]; then
+        echo "[-] Unable to find Python, is it installed?" 1>&2
+        exit 1
+    fi
+    if [[ ! `which cargo` ]]; then
+        echo "[-] Unable to find cargo (Rust), is it installed?" 1>&2
+        exit 1
+    fi
 fi
 
 # Build Capstone
 
 echo "[*] Building Capstone ..."
 cd libs/capstone
-export CAPSTONE_ARCHS="x86"
-make -j12  || exit 1
-cd ../..
+if [[ "$is_windows" == "y" ]]; then
+    mkdir -p build
+    cd build
+    cmake -DCMAKE_BUILD_TYPE=Release -DCAPSTONE_BUILD_STATIC_RUNTIME=ON -DCAPSTONE_ARCHITECTURE_DEFAULT=OFF -DCAPSTONE_X86_SUPPORT=ON -G "NMake Makefiles" ..
+    nmake
+    cd ../../..
+else
+    export CAPSTONE_ARCHS="x86"
+    $make
+    cd ../..
+fi
 
 # Build Intel XED
 
 echo "[*] Building Intel XED ..."
 cd libs/intelxed
-python mfile.py --opt=3
+$python mfile.py --opt=3
 cd ../..
 
 # Build DiStorm
 
 echo "[*] Building DiStorm ..."
-if [[ `uname -s` == "Darwin" ]]; then 
+if [[ "$is_mac" == "y" ]]; then 
     distorm_subdir=mac
     distorm_bin=libdistorm3.dylib
+elif [[ "$is_windows" == "y" ]]; then
+    distorm_subdir=win32
 else
     distorm_subdir=linux
     distorm_bin=libdistorm3.so.3.4.0
 fi
 cd libs/distorm/make/${distorm_subdir}
-make || exit 1
-echo $(pwd)
-cp ${distorm_bin} ../../../../bench/distorm/
+if [[ "$is_windows" == "y" ]]; then
+    msbuild.exe cdistorm.vcxproj -p:Configuration=clib -p:Platform=x64
+else
+    $make
+    cp ${distorm_bin} ../../../../bench/distorm/
+fi
 cd ../../../..
-if ! [[ `uname -s` == "Darwin" ]]; then
+if [[ "$is_linux" == "y" ]]; then
     cd bench/distorm/
-    ln -s libdistorm3.so.3.4.0 libdistorm3.so.3
+    ln -s libdistorm3.so.3.4.0 libdistorm3.so.3 || true
     cd ../..
 fi
 
 # Build bddisasm
 
 echo "[*] Building bddisasm ..."
 cd libs/bddisasm
-make
+if [[ "$is_windows" == "y" ]]; then
+    msbuild.exe bddisasm/bddisasm.vcxproj -p:Configuration=Release -p:Platform=x64
+else
+    $make
+fi
 cd ../..
 
 # Build benchmark tools
 
 echo "[*] Building Capstone benchmark ..."
 cd bench/cs
-make || exit 1
+$make
```

Comment:
What is the intuition about removing the early exit on failure here (and in various other locations)? This forces you to manually scroll through a large amount of output flooded by succeeding builds to look for the failing ones, possibly not even realizing that some tools didn't build.

Reply:
`set -e` at the top of the file should cause an early exit on failure.

Reply:
Ah, neat, TIL!